### PR TITLE
[MM-57564] Removing notification blocking getUserProfiles call from notification function

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
- Previously we were making a call to get user profile irrespective of our notification preference, so the related code was moved down where we are constructing the username
- Even after moving the call down, it was observed that to get user profile, we were blocking the notification until we get the result from the api, which is not ideal and we would want to show the notification right away, and hence the call to user profile was eventually removed.
- The websocket event also sends the ’sender_name` in the body, which is also used as a fallback if the user is not there in the store.
- It also observed that the number of calls to user profile only due to this change have bought down the number of requests count by 5%, when app was running with 50K logged in user who were posting at the rate of 35 posts per second for the duration of 2 mins after app was running for 2mins.
CleanShot 2024-04-02 at 00 19 11@2x
- The average number of request was also brought down following the above same condition.
CleanShot 2024-04-02 at 00 21 44@2x

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57564

#### Screenshots
n/a

#### Release Note

```release-note
Unblocks notification call by using username instead of Full name in case of missing user profile.
```

### CODE REVIEW

**Review by agnivade — APPROVED** [review]
Wow this is a great find!

**Comment by hmhealey** [comment]
This will cause a change where, if the web app hasn't loaded a user who sends you a message, it could cause the notification to show that user's username instead of their display name because we're no longer loading that user's name. 

Perhaps, instead of changing this to never load that user, we could move this logic around so that we only fetch that user once we know that this notification will actually trigger a desktop notification so that it triggers less frequently? As-is, this function mixes logic for determining if a notification should be sent with logic for constructing the notification, so perhaps perhaps deferring this until after we determine whether or not to send the  if we deferred constructing the notification until as late as possible.

That said, I just reviewed https://github.com/mattermost/mattermost/pull/26604 which heavily modifies this function as well, so any bigger changes here will conflict with @devinbinnie's PR which is higher priority to get in than this one

**Comment by hmhealey** [comment]
Actually, if we change that, I just noticed that we also attempt to get the author of the post on [this line](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/actions/websocket_actions.jsx#L706), so one of these calls is redundant. Unfortunately, these are likely racing against each other and we don't have a good way of ensuring that that only happens once between those two calls

**Comment by agnivade** [comment]
> That said, I just reviewed https://github.com/mattermost/mattermost/pull/26604 which heavily modifies this function as well, so any bigger changes here will conflict with @devinbinnie's PR which is higher priority to get in than this one

0/5 on the end result, but priority wise, both should be equal.

**Comment by devinbinnie** [comment]
> This will cause a change where, if the web app hasn't loaded a user who sends you a message, it could cause the notification to show that user's username instead of their display name because we're no longer loading that user's name.
> 
> Perhaps, instead of changing this to never load that user, we could move this logic around so that we only fetch that user once we know that this notification will actually trigger a desktop notification so that it triggers less frequently? As-is, this function mixes logic for determining if a notification should be sent with logic for constructing the notification, so perhaps perhaps deferring this until after we determine whether or not to send the if we deferred constructing the notification until as late as possible.
> 
> That said, I just reviewed #26604 which heavily modifies this function as well, so any bigger changes here will conflict with @devinbinnie's PR which is higher priority to get in than this one

@M-ZubairAhmed's change can go in first, it's much simpler and doesn't really touch the areas I was changing anyways.

**Comment by M-ZubairAhmed** [comment]
> This will cause a change where, if the web app hasn't loaded a user who sends you a message, it could cause the notification to show that user's username instead of their display name because we're no longer loading that user's name.
> 
> Perhaps, instead of changing this to never load that user, we could move this logic around so that we only fetch that user once we know that this notification will actually trigger a desktop notification so that it triggers less frequently? As-is, this function mixes logic for determining if a notification should be sent with logic for constructing the notification, so perhaps perhaps deferring this until after we determine whether or not to send the if we deferred constructing the notification until as late as possible.
> 
> That said, I just reviewed #26604 which heavily modifies this function as well, so any bigger changes here will conflict with @devinbinnie's PR which is higher priority to get in than this one

I've considered moving this piece of code further down in the action logic previously, but I've decided against it for two main reasons. 
Firstly, when we already have the message and user ID, it doesn't seem justifiable to delay the notification, even for the brief duration it takes to request the user profile. Prioritizing the display of notifications with all available information, including the message and user ID, is essential.
Secondly, the absence of the username is typically a result of a race condition, as the profiles for users in channels are usually loaded during the initial load. While it's a minor inconvenience to display the user ID instead of the user's display name due to this race condition, I believe the small probability of this occurring makes it an acceptable compromise.
In summary, while acknowledging the slight imperfection caused by this race condition, the benefits of ensuring timely notifications with the most critical information outweigh the drawbacks.

**Comment by hmhealey** [comment]
We're somewhat sacrificing UX for performance here, although I'm not sure how much it's likely to affect either of those since this should only come up pretty rarely. I'm not opposed to that though since I feel like we tend to sacrifice performance for UX, and either way, this is likely to not happen too much. Perhaps we could get around that by sending down the sender's full name and nickname similar to `sender_name`, but I don't think we need to worry about that now. I think John should be aware of this though

> Secondly, the absence of the username is typically a result of a race condition, as the profiles for users in channels are usually loaded during the initial load.

It's also possible that this is the first time a user has posted in a channel or in a thread that you've seen, so it's not just a race condition case, but I imagine the chance of that happening will be very low.

---

@johndavidlugtu Just a heads up that this might lead to you getting a desktop notification that is sent from a user's username instead of their full name, but that'll only happen if you've never seen that user or a post/thread of theirs as described above. It will ensure that those notifications aren't delayed to load the other user though

**Review by hmhealey — APPROVED** [review]

**Comment by marianunez** [comment]
> Just a heads up that this might lead to you getting a desktop notification that is sent from a user's username instead of their full name, but that'll only happen if you've never seen that user or a post/thread of theirs as described above. It will ensure that those notifications aren't delayed to load the other user though

+1 on prioritizing performance over this small UX degradation that will not happen in all notifications. 

@esethna heads up on this change of behavior on notifications.

**Review by marianunez — APPROVED** [review]
Thanks @M-ZubairAhmed!

**Review by yasserfaraazkhan — APPROVED** [review]
notifications are recieved as expected on browser and desktop.

**Comment by yasserfaraazkhan** [comment]
Tested this locally with desktop app and browser. Notifications we sent as expected.
